src/forcing/ForcingsEngineLumpedDataProvider.cpp

```diff
@@ -0,0 +1,259 @@
+                               std::to_string(shape[1]) + ", " +
+                               std::to_string(shape[2]);
+
+        const auto index_str = (d_idx == bad_index ? "?" : std::to_string(d_idx)) + ", " +
+                               (v_idx == bad_index ? "?" : std::to_string(v_idx));
+
+        throw std::out_of_range{
+            "Failed to get ForcingsEngineLumpedDataProvider value at index {" + index_str + "}" +
+            "\n  Shape    : {" + shape_str + "}" +
+            "\n  Divide   : `" + divide_id + "`" +
+            "\n  Variable : `" + variable  + "`"
+        };
+    }
+
+    return at(d_idx, v_idx, previous);
+}
+
+double Provider::at(
+    std::size_t divide_idx,
+    std::size_t variable_idx,
+    bool      previous
+)
+{
+    if (divide_idx == bad_index || variable_idx == bad_index) {
+        const auto shape = var_cache_.shape();
+        const auto shape_str = std::to_string(shape[0]) + ", " +
+                               std::to_string(shape[1]) + ", " +
+                               std::to_string(shape[2]);
+
+        const auto index_str = (divide_idx == bad_index ? "?" : std::to_string(divide_idx)) + ", " +
+                               (variable_idx == bad_index ? "?" : std::to_string(variable_idx));
+
+        throw std::out_of_range{
+            "Failed to get ForcingsEngineLumpedDataProvider value at index {" + index_str + "}" +
+            "\n  Shape          : {" + shape_str + "}"
+        };
+    }
+    
+    return var_cache_.at({{previous ? 1U : 0U, divide_idx, variable_idx}});
+}
+
+double Provider::get_value(
+    const CatchmentAggrDataSelector& selector,
+    ReSampleMethod m
+)
+{
+    const auto start = clock_type::from_time_t(selector.get_init_time());
+    const auto end = std::chrono::seconds{selector.get_duration_secs()} + start;
+    const auto step = std::chrono::seconds{this->record_duration()};
+    const std::string id = selector.get_id();
+    const std::string var = selector.get_variable_name();
+
+    if (m == ReSampleMethod::SUM || m == ReSampleMethod::MEAN) {
+        double acc = 0.0;
+        auto current_time = start;
+        while (current_time < end) {
+            if (!this->next()) {
+                break;
+            }
+
+            acc += this->at(id, var);
+            current_time += step;
+        }
+
+        if (m == ReSampleMethod::MEAN) {
+            const auto time_step_seconds = step.count();
+            const auto time_duration = std::chrono::duration_cast<std::chrono::seconds>(current_time.time_since_epoch()).count();
+            const auto num_time_steps = time_duration / time_step_seconds;
+            acc /= num_time_steps;
+        }
+
+        return acc;
+    }
+
+    throw std::runtime_error{"Given ReSampleMethod " + std::to_string(m) + " not implemented."};
+}
+
+std::vector<double> Provider::get_values(
+    const CatchmentAggrDataSelector& selector,
+    ReSampleMethod m
```

Comment:
Does this need to constrain acceptable values of `ReSampleMethod`, or gain the same functionality as the scalar case?

Reply:
I'm not sure to be honest. This one is tricky for me because I don't think resampling makes much sense in context of the `get_values` function -- at least, not for `CatchmentAggrDataSelector`, maybe for gridded data selectors with a target grid spec -- so I'm not sure what that would look like in this case.

I think my understanding of the implementation was that `get_values` gets all the values for that variable & catchment at each timestep in the given selector duration, so no resampling can really be done -- however, with the `get_value` function, it's resampled across the duration, so it makes sense there. I may be missing something though...